### ISSUE TITLE
Migration fails in 7.10.2 when trying to add unique index on tags

### ISSUE THREAD

**Issue by dreis2211** [gitlabhq/gitlabhq#9255]
Hey,

i just updated to 7-10-stable and experienced a bug when trying to migrate our database.

```
gitlab:/home/git/gitlab# sudo -u git -H bundle exec rake db:migrate RAILS_ENV=production
== 20150425164648 AddMissingUniqueIndices: migrating ==========================
-- add_index(:tags, :name, {:unique=>true})
rake aborted!
StandardError: An error has occurred, all later migrations canceled:

Mysql2::Error: Duplicate entry 'ios' for key 'index_tags_on_name': CREATE UNIQUE INDEX `index_tags_on_name`  ON `tags` (`name`) /home/git/gitlab/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.9/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:303:in `query'
/home/git/gitlab/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.9/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:303:in `block in execute'
/home/git/gitlab/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.9/lib/active_record/connection_adapters/abstract_adapter.rb:373:in `block in log'
/home/git/gitlab/vendor/bundle/ruby/2.1.0/gems/activesupport-4.1.9/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
/home/git/gitlab/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.9/lib/active_record/connection_adapters/abstract_adapter.rb:367:in `log'
/home/git/gitlab/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.9/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:303:in `execute'
/home/git/gitlab/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.9/lib/active_record/connection_adapters/mysql2_adapter.rb:228:in `execute'
/home/git/gitlab/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.9/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:525:in `add_index'
/home/git/gitlab/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.9/lib/active_record/migration.rb:656:in `block in method_missing'
/home/git/gitlab/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.9/lib/active_record/migration.rb:628:in `block in say_with_time'
/home/git/gitlab/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.9/lib/active_record/migration.rb:628:in `say_with_time'
/home/git/gitlab/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.9/lib/active_record/migration.rb:648:in `method_missing'
/home/git/gitlab/vendor/bundle/ruby/2.1.0/gems/activerecord-4.1.9/lib/active_record/migration.rb:413:in `method_missing'
/home/git/gitlab/db/migrate/20150425164648_add_missing_unique_indices.acts_as_taggable_on_engine.rb:4:in `up'
```

The problem is that unique indexes in MySQL are depending on the underlying collation of the field. In GitLab this is case-insensitive as default, so adding unique indexes will break if there is a row with "IOS" and one with "ios" in the tags table.

The commit that introduced the bug, is https://github.com/gitlabhq/gitlabhq/commit/81634bc1f7d8abb1686ca0cf2318bc02f3c7d97c

It already contains a change of the collation, but is applied too late.

I don't know what procedure you're following in that case. Deleting the migration, adjusting it, adjusting the table or whatever. Otherwise I'd do the PR myself.

See the following commit for additional information, that tags should be strictly case matching.
https://github.com/gitlabhq/gitlabhq/commit/45de7d263d8494d0aaaf99d7a6c44f7f9d2adf09

Cheers.

**Comment by drmonty** [comment]
Same issue occurred here on migrating from 7.10.1 -> 7.10.2

**Comment by stefan-it** [comment]
I can confirm this issue. I guess it is a regression introduced with this [Commit](https://github.com/gitlabhq/gitlabhq/commit/81634bc1f7d8abb1686ca0cf2318bc02f3c7d97c).

**Comment by dreis2211** [comment]
Thanks @stefan-it .

Yes, it is - I already referenced it ;-)

**Comment by dreis2211** [comment]
It would work if the collation change in 20150425164651_change_collation_for_tag_names.acts_as_taggable_on_engine.rb would be applied before the unique index is added.

**Comment by cosmix** [comment]
@dreis2211, are you sure? It seems to me that it's about uniqueness per project, independently of collation/case. Specifically, the addition of a unique index would fail in exactly the same fashion, if the same label/tag is used across different groups/projects, even if a less restrictive collation is set.

The `index_tags_on_name` doesn't take into account the `project` field at all. It should.

**Comment by dreis2211** [comment]
@cosmix, I'm not sure if I get your point.

### Case 1:
Assuming we have one tag **iOS** with the ID 1 in the `tags` table and we have projects with the IDs 1 and 2, we will have the following sample `taggings`table:

tag_id | taggable_id
------------ | -------------
1 | 1
1 | 2

**No problem so far with an unique index.**

### Case 2:
Assuming we have one tag **iOS** with the ID 1 in the `tags` table and we have projects with the IDs 1 and 2, we will have the following sample `taggings`table:

tag_id | taggable_id
------------ | -------------
1 | 1
1 | 2
2 | 1
2 | 2

**Also no problem from the unique index point of view.**

It might be bonkers from the usability point of view, but as mentioned in https://github.com/gitlabhq/gitlabhq/commit/45de7d263d8494d0aaaf99d7a6c44f7f9d2adf09 labels/tags should be different.

I actually don't care what the usability should be here - that's for others to decide. And might be decided at a later point in time. The point of this issue is no question of usability for us at the moment, but simply a migration failure, that could be solved by applying the collation migration before the unique index is created.

Cheers

**Comment by cosmix** [comment]
Indeed, I was referring to the migration not the usability consequences of the current design, but, alas, you're right. I was experiencing the same migration failure and --- erroneously --- looking at the `labels` table.

**Comment by Razer6** [comment]
Also mentioned in https://gitlab.com/gitlab-org/gitlab-ce/issues/1589

**Comment by dreis2211** [comment]
Thanks @Razer6 - was about to link that as well ;)

**Comment by vsizov** [comment]
fixed here https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/623

**Comment by dreis2211** [comment]
@vsizov As mentioned in there, this is a problem for peope expecting case-sensitive labels. Like it was added in commit 45de7d2

**Comment by stanhu** [comment]
MySQL users need this as well: https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/625

**Comment by JobV** [comment]
Thanks everyone for reporting. We're working on fixing the issues with the migrations.

**Comment by vsizov** [comment]
fixed by f6c0ef5553fde03ae3a04ba48733455a5415a3ac
